rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeper.java

```diff
@@ -296,6 +297,43 @@ public void onRemoval(
                 .build();
     }
 
+    @Override
+   public boolean readData(String path, long offset, int length, long fileSize, long lastModified, int clusterType)
+           throws TException
+   {
+       int blockSize = CacheConfig.getBlockSize(conf);
+       byte[] buffer = new byte[blockSize];
+       BookKeeperFactory bookKeeperFactory = new BookKeeperFactory(this);
+       CachingNativeS3FileSystem fs = null;
+       FSDataInputStream inputStream = null;
+       long startBlock = offset / blockSize;
+       long endBlock = ((offset + (length - 1)) / CacheConfig.getBlockSize(conf)) + 1;
+       try {
+           int idx = 0;
+           List<BlockLocation> blockLocations = getCacheStatus(path, fileSize, lastModified, startBlock, endBlock, clusterType);
+
+           for (int blockNum = (int) startBlock; blockNum < endBlock; blockNum++, idx++) {
+               int readStart = blockNum * blockSize;
+               if (blockLocations.get(idx).getLocation() == Location.LOCAL) {
+                   if (fs == null) {
+                       fs = new CachingNativeS3FileSystem(bookKeeperFactory, new Path(path), conf);
+                       inputStream = fs.open(new Path(path), blockSize);
+                   }
+                   inputStream.seek(readStart);
```

Comment:
There is a potential optimization to speed up warmup. Right now the work of getCacheStatus is repeated, once in this method and once in inputStream.read. We can rather create a RemoteReadRequestChain here directly and call add(), call() and updateCacheStatus() methods here rather than inputStream.read()